pkg/auth/providers/ldap/ldap_client_test.go

```diff
@@ -1,400 +1,561 @@
 package ldap
 
 import (
-	"context"
-	"crypto/tls"
-	"crypto/x509"
-	"reflect"
+	"fmt"
 	"testing"
-	"time"
 
 	ldapv3 "github.com/go-ldap/ldap/v3"
-	"github.com/pkg/errors"
-	"github.com/rancher/norman/types"
-	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/norman/httperror"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	ldapFakes "github.com/rancher/rancher/pkg/auth/providers/common/ldap"
 	"github.com/rancher/rancher/pkg/auth/tokens"
-	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
-	"github.com/rancher/rancher/pkg/user"
-	wcorev1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	apitypes "k8s.io/apimachinery/pkg/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 const (
-	DummySAUsername     = "satestuser1"
-	DummySAUPassword    = "sapassword123"
-	UserObjectClassName = "inetOrgPerson"
+	saDN                = "cn=sa,ou=users,dc=foo,dc=bar"
+	saPassword          = "secret"
+	userName            = "user"
+	userDN              = "cn=user,ou=users,dc=foo,dc=bar"
+	userPassword        = "secret"
+	userObjectClassName = "inetOrgPerson"
 )
 
-func Test_ldapProvider_loginUser(t *testing.T) {
-	type fields struct {
-		ctx                   context.Context
-		authConfigs           v3.AuthConfigInterface
-		secrets               wcorev1.SecretController
-		userMGR               mockUserManager
-		tokenMGR              *tokens.Manager
-		certs                 string
-		caPool                *x509.CertPool
-		providerName          string
-		testAndApplyInputType string
-		userScope             string
-		groupScope            string
+func TestLDAPProviderLoginUser(t *testing.T) {
```

Comment:
Deliberately break out table-driven tests into subtests in order to be able to easily run/debug individual test cases.